perma_web/perma/tasks.py

```diff
@@ -302,586 +704,331 @@ def warn_on_exception(message="Exception in block:", exception_type=Exception):
-                break
-
-            # Give up if downloaded size exceeds MAX_ARCHIVE_FILE_SIZE.
-            # Amount captured so far is the sum of the bytes recorded by warcprox,
-            # and the bytes pending in our background threads.
-            capture_size = sum(getattr(thread, 'pending_data', 0) for thread in thread_list) + \
-                           sum(proxied_response.response_recorder.len for proxied_response in proxied_responses)
-            if capture_size > settings.MAX_ARCHIVE_FILE_SIZE:
-                print "Halting for size"
-                break
-
-            # Show progress to user
-            progress = int(progress) + wait_time/AFTER_LOAD_TIMEOUT
-            display_progress(progress, "Waiting for post-load requests")
-
-            # Sleep and update our list
-            time.sleep(.5)
-            unfinished_proxied_pairs = [pair for pair in unfinished_proxied_pairs if not pair[1]]
+                safe_save_fields(link.screenshot_capture, status='failed')
 
-        # screenshot capture
-        if have_html and browser_still_running(browser):
-            # get page size to decide whether to take a screenshot
-            capture_screenshot = False
-            pixel_count = 0
-            try:
-                root_element = browser.find_element_by_tag_name('body')
-            except (NoSuchElementException, URLError):
-                try:
-                    root_element = browser.find_element_by_tag_name('frameset')
-                except (NoSuchElementException, URLError):
-                    # If we can't find the root element, we just won't capture a screenshot.
-                    # NoSuchElementException: HTML structure is weird somehow.
-                    # URLError: the headless browser has gone away for some reason.
-                    root_element = None
-            if root_element:
-                page_size = root_element.size
-                pixel_count = page_size['width'] * page_size['height']
-                capture_screenshot = pixel_count < settings.MAX_IMAGE_SIZE
-
-            # take screenshot after all requests done
-            if capture_screenshot:
-                progress += 1
-                display_progress(progress, "Taking screenshot")
-                screenshot_data = browser.get_screenshot_as_png()
-                link.screenshot_capture.write_warc_resource_record(screenshot_data)
-                save_fields(link.screenshot_capture, status='success')
-            else:
-                print "Not saving screenshots! %s" % ("Page size is %s pixels." % pixel_count if pixel_count else "")
-                save_fields(link.screenshot_capture, status='failed')
-        else:
-            # no screenshot if not HTML
-            save_fields(link.screenshot_capture, status='failed')
+        except HaltCaptureException:
+            print("HaltCaptureException thrown")
+    except SoftTimeLimitExceeded:
```

Comment:
Yes.... If the tempfile/memory situation allows, I'm wondering if the right thing to so is to pass the file I/O off to a separate task, one that celery can keep retrying if (for instance) some network situation temporarily makes it impossible to save to s3.